dask/blockwise.py

```diff
@@ -900,46 +805,48 @@ def make_blockwise_graph(
     output_blocks = output_blocks or list(
         itertools.product(*[range(dims[i]) for i in out_indices])
     )
+    from dask._task_spec import DataNode
 
     dsk = {}
-    # Create argument lists
     for out_coords in output_blocks:
-        deps = set()
+        this_task = task
         coords = out_coords + dummies
         args = []
-        for cmap, axes, (arg, ind) in zip(coord_maps, concat_axes, argpairs):
+        for cmap, axes, (arg, ind), key in zip(
+            coord_maps, concat_axes, argpairs, keys, strict=True
+        ):
+            if key not in task.dependencies:
+                # FIXME: This feels like a bug
+                continue
             if ind is None:
-                args.append(arg)
+                if not isinstance(arg, (GraphNode, TaskRef)):
+                    this_task = this_task.substitute({key: DataNode(None, arg)})
+                else:
+                    this_task = this_task.substitute({key: arg})
+                continue
+            arg_coords = tuple(coords[c] for c in cmap)
+            if arg in io_deps:
+                val = parse_input(io_deps[arg].get(arg_coords, arg_coords))
+                if not isinstance(val, GraphNode):
+                    val = DataNode(None, val)
+                this_task = this_task.substitute({key: val})
             else:
-                arg_coords = tuple(coords[c] for c in cmap)
+                subs = {}
                 if axes:
-                    tups = lol_product((arg,), arg_coords)
-                    if arg not in io_deps:
-                        deps.update(flatten(tups))
-
+                    tups = lol_product((arg,), arg_coords, as_taskref=True)
                     if concatenate:
-                        tups = (concatenate, tups, axes)
+                        tups = Task(key, concatenate, tups, axes)
+                    subs[key] = tups
                 else:
-                    tups = (arg,) + arg_coords
-                    if arg not in io_deps:
-                        deps.add(tups)
-                # Replace "place-holder" IO keys with "real" args
-                if arg in io_deps:
-                    # We don't want to stringify keys for args
-                    # we are replacing here
-                    idx = tups[1:]
-                    args.append(io_deps[arg].get(idx, idx))
-                else:
-                    args.append(tups)
-        out_key = (output,) + out_coords
-
-        args.insert(0, func)
-        dsk[out_key] = tuple(args)
-
+                    subs[key] = (arg, *arg_coords)
+                this_task = this_task.substitute(subs)
+        new_key = (output,) + out_coords
+        assert isinstance(this_task, Task)
+        dsk[new_key] = Task.fuse(this_task, *args, key=new_key)
     return dsk
 
 
-def lol_product(head, values):
+def lol_product(head, values, as_taskref=False):
```

Comment:
Should we document `as_taskref`?

Reply:
I don't have interest in documenting this function properly. This is an internal utility. I changed the function name to lead with an underscore.